### PR TITLE
add jwks guide

### PR DESCRIPTION
**Pull request by sdelamo**

### CODE REVIEW

File: guides/micronaut-security-keys-jwks/metadata.json [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  "authors": ["Sergio del Amo"],
+  "tags": ["jwk", "graalvm"],
+  "category": "AWS",
+  "publicationDate": "2021-05-27",
+  "languages": ["java"],
+  "testFramework": "junit",
+  "apps": [
+    {
+      "name": "default",
+      "features": ["security-jwt"]
```
Comment (ilopmar) [inline guides/micronaut-security-keys-jwks/metadata.json]: Remove `graalvm` from `tags` and add it to `features`

File: guides/micronaut-security-keys-jwks/java/src/main/java/example/micronaut/Application.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package example.micronaut;
+
+import io.micronaut.runtime.Micronaut;
+
+public class Application {
```
Comment (ilopmar) [inline guides/micronaut-security-keys-jwks/java/src/main/java/example/micronaut/Application.java]: This is the generated class, please remove it from the guide.

File: guides/micronaut-security-keys-jwks/java/src/main/resources/application.yml [hunk 3]
```diff
@@ -0,0 +1,3 @@
+micronaut:
+  application:
+    name: micronautguide
```
Comment (ilopmar) [inline guides/micronaut-security-keys-jwks/java/src/main/resources/application.yml:3]: Default generated file. Please remove it.

File: guides/micronaut-security-keys-jwks/java/src/main/resources/logback.xml [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+             ch.qos.logback.classic.encoder.PatternLayoutEncoder by default -->
+        <encoder>
+            <pattern>%cyan(%d{HH:mm:ss.SSS}) %gray([%thread]) %highlight(%-5level) %magenta(%logger{36}) - %msg%n</pattern>
+        </encoder>
+    </appender>
+
+    <root level="info">
+        <appender-ref ref="STDOUT" />
+    </root>
+</configuration>
```
Comment (ilopmar) [inline guides/micronaut-security-keys-jwks/java/src/main/resources/logback.xml]: Default generated file. Please remove it.

File: guides/micronaut-security-keys-jwks/micronaut-security-keys-jwks.adoc [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+Create a refreshable bean to load the secondary key:
+
+source:SecondarySignatureConfiguration[]
+
+<1> `@Refreshable` scope is a custom scope that allows a bean’s state to be refreshed via the https://docs.micronaut.io/latest/guide/index.html#refreshEndpoint[/refresh] endpoint.
+
+=== Primary Key
+
+Create an abstract class which extends from `AbstractRSASignatureConfiguration` but also implements `AbstractRSASignatureConfiguration`.
```
Comment (ilopmar) [inline guides/micronaut-security-keys-jwks/micronaut-security-keys-jwks.adoc]: Typo: "...but also implements `RSASignatureGeneratorConfiguration`", not `AbstractRSASignatureConfiguration`

File: guides/micronaut-security-keys-jwks/micronaut-security-keys-jwks.adoc [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+source:PrimarySignatureConfiguration[]
+
+<1> `@Refreshable` scope is a custom scope that allows a bean’s state to be refreshed via the https://docs.micronaut.io/latest/guide/index.html#refreshEndpoint[/refresh] endpoint.
+<2> You need to qualify with @Named `generator` the signature configuration bean which you wish to use to sign your JSON web tokens.
+
+=== JWKProvider
+
+To enable the https://micronaut-projects.github.io/micronaut-security/latest/guide/#keys[Keys Controller] you have to provide at least a bean of type: `JwkProvider`.
+
+source:JSONWebKeysProvider[]
```
Comment (ilopmar) [inline guides/micronaut-security-keys-jwks/micronaut-security-keys-jwks.adoc]: Error rendering this file: "unresolver directive...."

File: guides/micronaut-security-keys-jwks/java/src/test/java/example/micronaut/TokenGeneratorTest.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        Assertions.assertDoesNotThrow(() -> JWTParser.parse(jwtString));
+
+        JWT jwt = JWTParser.parse(jwtString);
+
+        Assertions.assertTrue(jwt instanceof SignedJWT);
+
+        tokenValidator.validateToken(jwtString, null);
+
+        Assertions.assertEquals(1, jwt.getJWTClaimsSet().getClaims().keySet().size());
+        Assertions.assertEquals("sergio", jwt.getJWTClaimsSet().getSubject());
```
Comment (ilopmar) [inline guides/micronaut-security-keys-jwks/java/src/test/java/example/micronaut/TokenGeneratorTest.java]: Please use static imports for `Assertions`

File: guides/micronaut-security-keys-jwks/java/src/test/java/example/micronaut/ValidateSecondarySignedJwtTest.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+        Assertions.assertTrue(jwtOptional.isPresent());
+
+        String jwtString = jwtOptional.get();
+
+        Assertions.assertDoesNotThrow(() -> JWTParser.parse(jwtString));
+
+        JWT jwt = JWTParser.parse(jwtString);
+
+        Assertions.assertTrue(jwt instanceof SignedJWT);
```
Comment (ilopmar) [inline guides/micronaut-security-keys-jwks/java/src/test/java/example/micronaut/ValidateSecondarySignedJwtTest.java]: Same here, use static imports for assertions

File: guides/micronaut-security-keys-jwks/java/src/test/java/example/micronaut/KeysTest.java [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+@MicronautTest
+public class KeysTest {
+
+    @Inject
+    @Client("/")
+    HttpClient httpClient;
+
+    @Test
+    void whenYouDefineABeanOfTypeJwkProviderThenTheKeysEndpointIsExposed() {
+        Assertions.assertDoesNotThrow(() -> httpClient.toBlocking().exchange(HttpRequest.GET("/keys")));
```
Comment (ilopmar) [inline guides/micronaut-security-keys-jwks/java/src/test/java/example/micronaut/KeysTest.java]: Use static import for assertions

**Review by ilopmar — CHANGES_REQUESTED** [review]
Besides my specific comments one thing that I think we should explain is how to get the JWK primary and secondary keys that you use in the tests and when running the application.
For someone like me that doesn't know too much about security is not clear how I can generate those keys.

Another thing to improve the guide would be to show how to refresh the keys. Maybe something like: get the keys, send a request to `/refresh`, get the keys again and check that you get new keys.

File: guides/micronaut-security-keys-jwks/metadata.json (hunk 1 above)
Comment (sdelamo) [reply guides/micronaut-security-keys-jwks/metadata.json]: Done.

File: guides/micronaut-security-keys-jwks/java/src/main/java/example/micronaut/Application.java (hunk 2 above)
Comment (sdelamo) [reply guides/micronaut-security-keys-jwks/java/src/main/java/example/micronaut/Application.java]: deleted

File: guides/micronaut-security-keys-jwks/java/src/main/resources/logback.xml (hunk 4 above)
Comment (sdelamo) [reply guides/micronaut-security-keys-jwks/java/src/main/resources/logback.xml]: deleted

File: guides/micronaut-security-keys-jwks/micronaut-security-keys-jwks.adoc (hunk 6 above)
Comment (sdelamo) [reply guides/micronaut-security-keys-jwks/micronaut-security-keys-jwks.adoc]: I've replaced 

`source:JsonWebKeysProvider[]`

with: 

`source:JSONWebKeysProvider[]`

File: guides/micronaut-security-keys-jwks/java/src/test/java/example/micronaut/ValidateSecondarySignedJwtTest.java (hunk 8 above)
Comment (sdelamo) [reply guides/micronaut-security-keys-jwks/java/src/test/java/example/micronaut/ValidateSecondarySignedJwtTest.java]: I have replaced to use static imports.

File: guides/micronaut-security-keys-jwks/java/src/test/java/example/micronaut/KeysTest.java (hunk 9 above)
Comment (sdelamo) [reply guides/micronaut-security-keys-jwks/java/src/test/java/example/micronaut/KeysTest.java]: I've used static imports.

File: guides/micronaut-security-keys-jwks/java/src/test/java/example/micronaut/TokenGeneratorTest.java (hunk 7 above)
Comment (sdelamo) [reply guides/micronaut-security-keys-jwks/java/src/test/java/example/micronaut/TokenGeneratorTest.java]: I have used static imports

File: guides/micronaut-security-keys-jwks/micronaut-security-keys-jwks.adoc (hunk 5 above)
Comment (sdelamo) [reply guides/micronaut-security-keys-jwks/micronaut-security-keys-jwks.adoc]: Fixed.

**Comment by sdelamo** [comment]
> Besides my specific comments one thing that I think we should explain is how to get the JWK primary and secondary keys that you use in the tests and when running the application.
> For someone like me that doesn't know too much about security is not clear how I can generate those keys.

I have linked to the [JWK CLI Guide](https://guides.micronaut.io/latest/micronaut-cli-jwkgen.html) which shows how to generate a guide.

**Comment by sdelamo** [comment]
> Another thing to improve the guide would be to show how to refresh the keys. Maybe something like: get the keys, send a request to `/refresh`, get the keys again and check that you get new keys.

I have added a `/refresh` endpoint section. I have another guide coming which shows secret rotation with AWS Secrets Manager.

**Review by ilopmar — APPROVED** [review]

**Pull request merged**